### ISSUE TITLE
HybridApp Windows WPF Webview doesn't render.

### ISSUE THREAD

**Issue by Kukks** [xamarin/MobileBlazorBindings#196]
Running the default template for hybrid apps does not render the webview when using the Windows WPF project.

**Comment by Eilon** [comment]
Hi @Kukks have you checked out the pre-requisites listed here? https://docs.microsoft.com/mobile-blazor-bindings/get-started#prerequisites? Specifically, on Windows you must have the Microsoft Edge Canary channel installed. It's side-by-side with any other version of Edge and any other browser, so even if you normally use another browser, it won't interfere with anything.

Can you confirm that you have Edge Canary installed?

**Comment by Kukks** [comment]
Ah that's my issue for sure then, I should have read the docs
more carefully, thanks!

Is there a way to mitigate this without requiring the user to install a
specific browser? Can we bundle a specific webview component to ensure the
experience is consistent across all Windows users?

**Comment by Eilon** [comment]
Right now Mobile Blazor Bindings uses the new Edge WebView2 control to host web content, and that WebView2 control currently requires Edge Canary because it's in preview. But in the future what's in Edge Canary will go to "regular" Edge that just about everyone will have anyway (does Windows Update install it? I can't recall.), so users won't need to install a pre-release Edge Canary.

But if in the future there is some requirement for end users, we'll definitely want to have some logic that at a minimum detects the missing component at runtime and says "Hey you need to install XYZ" but maybe we would also have a way of packaging it all up.

Those are certainly questions for the future because until WebView2 is finalized, there's not much that we should do pre-emptively because it might not be necessary in the future.

**Comment by Kukks** [comment]
@Eilon  I installed Edge Canary and used the template again and it still did not load.

**Comment by Kukks** [comment]
I have Version 87.0.669.0 (Official build) canary (64-bit) and Version 86.0.622.38 (Official build) (64-bit) installed, with the non-canary one set as default.

**Comment by Eilon** [comment]
Thanks for the info I'll take a look. At the very least this might be a sign that we need better diagnostics.

**Comment by Eilon** [comment]
Let's try this:

1. Open the solution/project for the hybrid app
1. In Visual Studio, go to Debug / Windows / Exception Settings
1. Fully check the "Common Language Runtime Exceptions" option (not just a grayed out checkbox, but a proper check):
![image](https://user-images.githubusercontent.com/202643/95951280-ff137a00-0daa-11eb-9c34-42809ee70013.png)
1. Press <kbd>F5</kbd> to run the app with the debugger attached

And then see if any exception pops up.

**Comment by Eilon** [comment]
As a point of comparison, I just got a brand new laptop, ran a hybrid app without Edge Canary installed, and of course it failed (with an exception seen in VS, but hidden from normal view). Then I installed Edge Canary and it ran fine (reported as `Version 88.0.670.0 (Official build) canary (64-bit)`).

**Comment by Eilon** [comment]
And it seems I have Edge Canary 88.<something> and you have 87. That's probably the problem. Can you check if yours says it's up-to-date?

![image](https://user-images.githubusercontent.com/202643/95953100-36375a80-0dae-11eb-864a-e2bd2d66dde5.png)

**Comment by Kukks** [comment]
![image](https://user-images.githubusercontent.com/1818366/95956633-e2a12d00-0dfe-11eb-8436-57cd0064d430.png)
![image](https://user-images.githubusercontent.com/1818366/95956654-e8970e00-0dfe-11eb-9308-9d5a40e91415.png)

Same result.

I enabled all CLR exceptions to throw on Rider and nothing threw.

**Comment by Eilon** [comment]
I just noticed that I don't even see any native UI in that screenshot, so I'm thinking this isn't a hybrid-specific issue. Are there any changes to the app from the default template? Can you publish your code on GitHub so I can try to reproduce the issue?

**Comment by ErikApption** [comment]
Same issue here (with Canarie 88 installed) and I get
`System.DllNotFoundException
  HResult=0x80131524
  Message=Unable to load DLL 'WebView2Loader.dll' or one of its dependencies: The specified module could not be found. (0x8007007E)
  Source=Microsoft.Web.WebView2.Core
  StackTrace:
   at Microsoft.Web.WebView2.Core.CoreWebView2Environment.CreateCoreWebView2EnvironmentWithOptions(String browserExecutableFolder, String userDataFolder, ICoreWebView2EnvironmentOptions options, ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler environment_created_handler)
   at Microsoft.Web.WebView2.Core.CoreWebView2Environment.<CreateAsync>d__3.MoveNext()
`

**Comment by Kukks** [comment]
> I just noticed that I don't even see any native UI in that screenshot, so I'm thinking this isn't a hybrid-specific issue. Are there any changes to the app from the default template? Can you publish your code on GitHub so I can try to reproduce the issue?

Ah yeah, sorry. I stripped out the native elements as my plans to use this project involve only the BlazorWebView.

I just tried again on a fresh template install  with both `Version 86.0.622.48 (Official build) (64-bit) and ``Version** 88.0.676.0 (Official build) canary (64-bit)` installed. The non-canary one is the default browser on this device. Could it be that the webview is attempting to use the non-canary version?

**Comment by Kukks** [comment]
Ok, I've noticed some weird behavior actually: 
![](https://i.imgur.com/XaTkUOK.gif)

**Comment by Eilon** [comment]
@ErikApption / @Kukks - can you confirm that your MyApplication.Windows.csproj has this direct reference:

```xml
    <PackageReference Include="Microsoft.Web.WebView2" Version="0.9.538-prerelease" />
```

(It will also have this reference _transitively_ via the `Microsoft.MobileBlazorBindings.WebView.Windows` package, but that is insufficient.)

**Comment by Eilon** [comment]
Oh and as far as I know setting the default browser doesn't matter. You just need to have Edge Canary installed.

**Comment by Eilon** [comment]
I don't know if it'll help, but I have a PR out to update to a new version of the WebView2 SDK: https://github.com/xamarin/MobileBlazorBindings/pull/200

Maybe that will address this issue. It works for me before & after this change but who knows.

**Comment by Kukks** [comment]
> @ErikApption / @Kukks - can you confirm that your MyApplication.Windows.csproj has this direct reference:
> 
> ```
>     <PackageReference Include="Microsoft.Web.WebView2" Version="0.9.538-prerelease" />
> ```
> 
> (It will also have this reference _transitively_ via the `Microsoft.MobileBlazorBindings.WebView.Windows` package, but that is insufficient.)

Already there
```
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <UseWPF>true</UseWPF>
    <NoWarn>$(NoWarn);NU1701</NoWarn>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.MobileBlazorBindings.WebView.Windows" Version="0.4.74-preview" />
    <PackageReference Include="Microsoft.Web.WebView2" Version="0.9.538-prerelease" />
    <PackageReference Include="System.Text.Json" Version="4.7.2" />
    <PackageReference Include="Xamarin.Forms" Version="4.7.0.968" />
    <PackageReference Include="Xamarin.Forms.Platform.WPF" Version="4.7.0.968" NoWarn="NU1701" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\FirstBlazorHybridApp\FirstBlazorHybridApp.csproj" />
  </ItemGroup>

</Project>

```

**Comment by jspuij** [comment]
@ErikApption (And maybe @Kukks) The message:

`System.DllNotFoundException HResult=0x80131524 Message=Unable to load DLL 'WebView2Loader.dll' or one of its dependencies: The specified module could not be found. (0x8007007E) Source=Microsoft.Web.WebView2.Core StackTrace: at Microsoft.Web.WebView2.Core.CoreWebView2Environment.CreateCoreWebView2EnvironmentWithOptions(String browserExecutableFolder, String userDataFolder, ICoreWebView2EnvironmentOptions options, ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler environment_created_handler) at Microsoft.Web.WebView2.Core.CoreWebView2Environment.<CreateAsync>d__3.MoveNext()`

is indicating that an assembly (or something native it depends on) cannot be found. Now unfortunately .NET core does not support fusionw to create an assembly binding log, but it does support something similar (but by far not as useful as fusionw), by enabling a few environment variables:

https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/default-probing

```
COREHOST_TRACE=1
```

and

```
COREHOST_TRACEFILE=<path>
```

This log might hold the key as to why Edge is not loading on your machines. If you could share it with us, that would be a great help.

**Comment by Kukks** [comment]
Here is the trace:
https://gist.github.com/Kukks/e955884b8bc12670a3d10b502ad25c22

**Comment by jspuij** [comment]
@Kukks I've had a look at the trace and unfortunately it only captures the managed part, and that looks fine. There is a native dll called WebView2Loader.dll that should be present in the binary output directory (you could check that), but if that is there tracability stops:

WebView2Loader is a native dll that uses OLE to communicate with Edge. There is no static or early binding inside that DLL:

https://i.imgur.com/z5EqG0H.png

We are relying on the Native dll to actually capture the error and forwarding it as an exception to the managed code. As long as you don't see any exception (even thrown but catched) it's hard to diagnose further.

**Comment by Kukks** [comment]
> @Kukks I've had a look at the trace and unfortunately it only captures the managed part, and that looks fine. There is a native dll called WebView2Loader.dll that should be present in the binary output directory (you could check that), but if that is there tracability stops:
> 
> WebView2Loader is a native dll that uses late binding to communicate with Edge. There is no static or early binding inside that DLL:
> 
> https://i.imgur.com/z5EqG0H.png
> 
> We are relying on the Native dll to actually capture the error and forwarding it as an exception to the managed code. As long as you don't see any exception (even thrown but catched) it's hard to diagnose further. I'm not even sure your issue is the same as the one reported by @ErikApption.

I can confirm that `WebView2Loader.dll` is present in the output dir.. 

One thing i've noticed is that once I  click the native increment button, the state updates but then the whole app stops working. The UI still seems to respond but clicking or interacting with the button does not work anymore (breakpoints on the click event don't work, no exceptions are triggered even with all the CLR exceptions option, nothing is logged). So perhaps there is more at play here? 

Is there a particular reason you are determined to use the Edge WebView? Microsoft branded browsers are prone to dev rage. :)

**Comment by jspuij** [comment]
@Kukks Unfortunately there isn't much else. CefSharp wraps an embedded chrome (which hugely increases executable size) and usually targets an outdated chrome version (3 to 4 releases behind). It also does not support .NET core. Finally its a non-microsoft dependency that MS would have to support (if this lives belong its expirimental phase).

I agree that this is not ideal and the bindings are a little rough. It has come a long way though (early versions were C++ only and the API was by no means stable). The default bundling of Edge with the recently released Oktober update should ensure that Edge is installed on most windows 10 devices in the coming year. I'm pretty sure we can target a stable API during the coming year as well.

Any chance that you could post your windows version as well?

**Comment by Kukks** [comment]
> @Kukks Unfortunately there isn't much else. CefSharp wraps an embedded chrome (which hugely increases executable size) and usually targets an outdated chrome version (3 to 4 releases behind). It also does not support .NET core. Finally its a non-microsoft dependency that MS would have to support (if this lives belong its expirimental phase).
> 
> I agree that this is not ideal and the bindings are a little rough. It has come a long way though (early versions were C++ only and the API was by no means stable). The default bundling of Edge with the recently released Oktober update should ensure that Edge is installed on most windows 10 devices in the coming year. I'm pretty sure we can target a stable API during the coming year as well.
> 
> Any chance that you could post your windows version as well? 


10 Pro 10.0.1362

**Comment by jspuij** [comment]
@Kukks I found this in the Webview2 feedback issue list (which is a good source anyway):

https://github.com/MicrosoftEdge/WebView2Feedback/issues/544

**Comment by Kukks** [comment]
> @Kukks I found this in the Webview2 feedback issue list (which is a good source anyway):
> 
> [MicrosoftEdge/WebView2Feedback#544](https://github.com/MicrosoftEdge/WebView2Feedback/issues/544)

Ok, tried the current beta, Dev and Canary edge editions.

Beta gives me a different result!
![](https://i.imgur.com/wuioq1u.gif)

**Comment by Kukks** [comment]
Beta is 87.0.664.12

**Comment by Kukks** [comment]
Dev also gives the same result as Beta
Dev is  Version 88.0.673.0 (Official build) dev (64-bit)

Noticed the following in the console:
`[Could not load assembly: Microsoft.MobileBlazorBindings.WebView.Windows, Version=0.4.0.0, Culture=neutral, PublicKeyToken=null for Attribute Xamarin.Forms.Platform.WPF.ExportRendererAttribute | Some render
ers may not be loaded] Registrar
`

**Comment by jspuij** [comment]
That one is pretty normal for a complete mismatch between WebView2 and its SDK. I’m afraid that your Canary edge does load, but has some incompatibility that requires an update of the SDK. Eilon has already updated to that SDK version. The update will be released with the next preview. Or you could clone the repo and build it yourself if you’re in a hurry.

**Comment by jspuij** [comment]
Ah, I see it has not been merged yet: https://github.com/xamarin/MobileBlazorBindings/pull/200

**Comment by jspuij** [comment]
I've just done a routine check with the current Canary, which is 88.0.677.0.
I've done:

```
dotnet new -i Microsoft.MobileBlazorBindings.Templates::0.4.74-preview
dotnet new blazorhybrid -o FirstBlazorHybridApp
```

Then I've opened the solution and pressed F5:

![image](https://user-images.githubusercontent.com/985974/96703222-e0a51400-1392-11eb-999d-1de7c97b7f92.png)

Pretty stumped ATM.

**Comment by Kukks** [comment]
> I've just done a routine check with the current Canary, which is 88.0.677.0.
> I've done:
> 
> ```
> dotnet new -i Microsoft.MobileBlazorBindings.Templates::0.4.74-preview
> dotnet new blazorhybrid -o FirstBlazorHybridApp
> ```
> 
> Then I've opened the solution and pressed F5:
> 
> ![image](https://user-images.githubusercontent.com/985974/96703222-e0a51400-1392-11eb-999d-1de7c97b7f92.png)
> 
> Pretty stumped ATM.

Those were pretty much my steps.


I just cloned the branch from #200  and still have the same issue >_<

**Comment by Kukks** [comment]
Ah wait, 
```
System.UnauthorizedAccessException: Access is denied. (0x80070005 (E_ACCESSDENIED))
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
   at Microsoft.Web.WebView2.Core.CoreWebView2Environment.CreateAsync(String browserExecutableFolder, String userDataFolder, CoreWebView2EnvironmentOptions options)
   at Microsoft.MobileBlazorBindings.WebView.Windows.WebViewExtendedAnaheimRenderer.HandleElementChangedAsync(ElementChangedEventArgs`1 e) in C:\Git\MobileBlazorBindings\src\Microsoft.MobileBlazorBindings.WebView.Windows\WebViewExtendedAnaheimRenderer.cs:line 64
```

**Comment by jspuij** [comment]
It tries to create a userdata folder in the current directory. Looks like that is not allowed?
The location of the userdata folder will be configurable in the future. But it isn't right now.

**Comment by Kukks** [comment]
!!
![image](https://user-images.githubusercontent.com/1818366/96704438-5fe71780-1394-11eb-890d-1db917bb2293.png)
It's the user data folder, I changed it manually and it worked. Unsure why the default does not work TBH

**Comment by jspuij** [comment]
I think it is because the current directory (inside rider) is set to the rider application folder.
Visual studio will change the working directory to the output folder during debug. This might very well be an issue with Rider that you would have not encountered with Visual Studio.

**Comment by Kukks** [comment]
In any case, I'm glad this is only an IDE issue. I was ready to revive my
Internet Explorer hate for Edge :)

**Comment by Kukks** [comment]
if @ErikApption can confirm that this is also his issue, I'll close this. Unless you prefer a new fresh issue

**Comment by jspuij** [comment]
Could you check https://www.jetbrains.com/help/rider/Run_Debug_Configuration_dotNet_Project.html?
Especially the working folder? I'm curious what's there in Rider.

**Comment by jspuij** [comment]
> 
> 
> if @ErikApption can confirm that this is also his issue, I'll close this. Unless you prefer a new fresh issue

Leave it open for now. I'm pretty sure his is different because of the different exception.

**Comment by Eilon** [comment]
Yeah I think @ErikApption 's issue is mostly likely different.

I just merged #200 to update to the latest WebView2 package and you can use the updated NuGet packages from the nightly feed, which you can read about here: https://docs.microsoft.com/mobile-blazor-bindings/contribute/nightly-builds

Maybe that fixes this odd issue, or maybe not 😄
